### PR TITLE
CTPPS: standard descriptions filled (9_3_X)

### PR DESCRIPTION
**Pull request by forthommel**
This PR populates the `fillDescriptions` methods for all `RecoCTPPS/TotemRPLocal` plugins to allow the automatic generation of the associated `cfi` files.
Work done prior to the foreseen HLT studies for CTPPS.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test workflow 136.772,136.780

**Comment by forthommel** [comment]
@cmsbuild seems tired :)

**Comment by slava77** [comment]
it's "busy" actually, waiting for available nodes to run


@smuzaffar 
is it justifiably busy or is there something going on?
I see only one PR test in progress
https://cmssdt.cern.ch/jenkins/job/ib-any-integration/

File: RecoCTPPS/TotemRPLocal/python/ctppsLocalTrackLiteProducer_cfi.py [hunk 1]
```diff
@@ -1,13 +1,12 @@
 import FWCore.ParameterSet.Config as cms
 
-ctppsLocalTrackLiteProducer = cms.EDProducer("CTPPSLocalTrackLiteProducer",
-    tagSiStripTrack = cms.InputTag("totemRPLocalTrackFitter"),
-    tagDiamondTrack = cms.InputTag("ctppsDiamondLocalTracks"),
+from RecoCTPPS.TotemRPLocal.ctppsLocalTrackLiteDefaultProducer_cfi import ctppsLocalTrackLiteDefaultProducer
```
Comment (perrotta) [inline RecoCTPPS/TotemRPLocal/python/ctppsLocalTrackLiteProducer_cfi.py]: This shouldn't be a cfi file, but a cff one.
Then you can import without the need of modifying the config name.

Of course,, if you do so the replacement  ctppsLocalTrackLiteProducer_cfi.py -> ctppsLocalTrackLiteProducer_cff.py should be ported in all configs which use it:

PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py 
RecoCTPPS/Configuration/python/recoCTPPS_cff.py
RecoCTPPS/TotemRPLocal/test/ctpps_reco_cfg.py

(see https://cmssdt.cern.ch/lxr/search?_filestring=&_string=psLocalTrackLiteProducer_cf )

File: RecoCTPPS/TotemRPLocal/plugins/CTPPSDiamondLocalTrackFitter.cc [hunk 2]
```diff
@@ -102,11 +102,41 @@ CTPPSDiamondLocalTrackFitter::produce( edm::Event& iEvent, const edm::EventSetup
...
+
+  edm::ParameterSetDescription trackingAlgoParams;
+  trackingAlgoParams.add<double>( "threshold", 1.5 );
+  trackingAlgoParams.add<double>( "thresholdFromMaximum", 0.5 );
+  trackingAlgoParams.add<double>( "resolution", 0.01 /* mm */ );
+  trackingAlgoParams.add<double>( "sigma", 0.1 );
+  trackingAlgoParams.add<double>( "startFromX", -0.5 /* mm */ );
+  trackingAlgoParams.add<double>( "stopAtX", 19.5 /* mm */ );
+
+  trackingAlgoParams.add<std::string>( "pixelEfficiencyFunction", "(TMath::Erf((x-[0]+0.5*[1])/([2]/4)+2)+1)*TMath::Erfc((x-[0]-0.5*[1])/([2]/4)-2)/4" );
```
Comment (forthommel) [inline RecoCTPPS/TotemRPLocal/plugins/CTPPSDiamondLocalTrackFitter.cc]: Hi @perrotta 
We would prefer to maintain the current high-flexibility definition of this efficiency function for now, as the diamond detector is still under heavy commissioning. Of course, as soon as this phase will be over, we will indeed switch to a more conventional way of defining these functions (i.e. even hardcoding this formula at the plugin level).

File: RecoCTPPS/TotemRPLocal/plugins/CTPPSLocalTrackLiteProducer.cc [hunk 3]
```diff
@@ -101,4 +102,18 @@ CTPPSLocalTrackLiteProducer::produce( edm::Event& iEvent, const edm::EventSetup&
...
+void
+CTPPSLocalTrackLiteProducer::fillDescriptions( edm::ConfigurationDescriptions& descr )
+{
+  edm::ParameterSetDescription desc;
+
+  desc.add<edm::InputTag>( "tagSiStripTrack", edm::InputTag( "totemRPLocalTrackFitter" ) );
+  desc.add<edm::InputTag>( "tagDiamondTrack", edm::InputTag( "ctppsDiamondLocalTracks" ) );
+  desc.add<bool>( "doNothing", true ); // disable the module by default
+
+  descr.add( "ctppsLocalTrackLiteDefaultProducer", desc );
```
Comment (forthommel) [inline RecoCTPPS/TotemRPLocal/plugins/CTPPSLocalTrackLiteProducer.cc]: This "Default" naming was introduced to generate this intermediate `cfi` as a "buffer" that is cloned (as `ctppsLocalTrackLiteProducer`), transformed, and inserted in the eras conditions [here](https://github.com/forthommel/cmssw/blob/45dc32c5ddfaa3a7a18b0192a5e79a9726ff6f06/RecoCTPPS/TotemRPLocal/python/ctppsLocalTrackLiteProducer_cfi.py).
Unless you have a better solution to have the eras feature add the `doNothing = cms.bool(False)` statement to his module when the `ctpps_2016` modifier is used?

File: RecoCTPPS/TotemRPLocal/plugins/TotemRPUVPatternFinder.cc [hunk 4]
```diff
@@ -296,4 +297,58 @@ void TotemRPUVPatternFinder::produce(edm::Event& event, const edm::EventSetup& e
...
+  edm::ParameterSetDescription exceptions_validator;
+  exceptions_validator.add<unsigned int>( "rpId" ); // RP id according to CTPPSDetId
+  exceptions_validator.add<unsigned int>( "minPlanesPerProjectionToFit_U" );
+  exceptions_validator.add<unsigned int>( "minPlanesPerProjectionToFit_V" );
+  exceptions_validator.add<double>( "threshold_U" );
+  exceptions_validator.add<double>( "threshold_V" );
+
+  std::vector<edm::ParameterSet> exceptions_default;
+
+  // if a RP or projection needs adjustment of the above settings, you can use the following format:
```
Comment (forthommel) [inline RecoCTPPS/TotemRPLocal/plugins/TotemRPUVPatternFinder.cc]: In b9f4ef0b26c66600112ec4d40dea97ffa5e7c830 the full CTPPS reconstruction example is amended to introduce (as comments) the usage of this set of exceptional settings to the U-V pattern finder.

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
abort test

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
@forthommel : thank you for attachng detailed comments to all variables: this is nice!

My comment was mostly referring to the (e.g.) \* mm *\ that were written in the code itself, and as such completely invisible to the user of the configuration file.

Anyhow, it looks fine now.

**Comment by forthommel** [comment]
@perrotta thanks a lot for your review! I have never been much of a `fillDescriptions` user, and am still slowly discovering it, so thanks for your useful advices :)
What do you think about this autogenerated `ctppsLocalTrackLiteDefaultProducer_cfi` and the eras integation?

**Comment by perrotta** [comment]
The important thing is that there is only one cfi that defines the  
configuration for a given plugin. It will then be used unambiguously  
for the ConfB parsing.

Then, if in the cff file which is actually accessed by the standard  
sequences you modify that config, or clone a new one out of it, that  
is not so relevant. There are examples of both in our code.

For what concerns your implementation, I am fine with it.

A.

Laurent Forthomme <notifications@github.com> ha scritto:

**Comment by perrotta** [comment]
please test
(last IB is broken but it should now get ignored for PR tests)

**Comment by perrotta** [comment]
+1
-    Automatically generated configurations are identical to the previous explicit ones
 -   No difference in the results of the jenkins tests, as expected
 -   As a further check, some 200 evts from the wf 133.78 were also run with and without this PR, and again there were no differences on the reco outputs

**Comment by davidlange6** [comment]
merge

**Pull request merged**